src/Connection.php

```diff
@@ -210,7 +219,23 @@ public function createExpressionBuilder(): ExpressionBuilder
     protected function connect(): DriverConnection
     {
         if ($this->_conn !== null) {
-            return $this->_conn;
+            $isTimeToCheck =  time() - $this->lastCheckedAt >= $this->heartbeat;
+            $noCheckNeeded = $this->heartbeat === null || $this->isChecking;
+
+            if ($noCheckNeeded || ! $isTimeToCheck) {
+                return $this->_conn;
+            }
+
+            $this->isChecking = true;
+
+            $isAvailable = $this->reconnectOnFailure();
+
+            $this->lastCheckedAt = time();
```

Comment:
Maybe could you store the previous call to `time()` in a variable to prevent doing two calls, also to prevent weird issues, likely be better to have a time older than the check than younger as in the current implementation.